File: sdk/npm/src/providers/email10min.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'email10min';
const BASE_URL = 'https://email10min.com';

const BROWSER_HEADERS: Record<string, string> = {
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
  'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
  'Cache-Control': 'no-cache',
  'DNT': '1',
  'Pragma': 'no-cache',
  'Upgrade-Insecure-Requests': '1',
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
  'sec-ch-ua': '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
};

const AJAX_HEADERS: Record<string, string> = {
  'Accept': 'application/json, text/plain, */*',
  'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
  'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
  'Origin': BASE_URL,
  'Referer': `${BASE_URL}/zh`,
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
  'X-Requested-With': 'XMLHttpRequest',
  'sec-ch-ua': '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'sec-fetch-dest': 'empty',
  'sec-fetch-mode': 'cors',
  'sec-fetch-site': 'same-origin',
};

/** 从响应头提取 `name=value` Cookie 串 */
function cookieHeaderFromResponse(headers: Headers): string {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  const lines = typeof h.getSetCookie === 'function' ? h.getSetCookie() : [];
  const parts: string[] = [];
  for (const line of lines) {
    const nv = line.split(';')[0].trim();
    if (nv) parts.push(nv);
  }
  if (!parts.length) {
    const single = headers.get('set-cookie');
    if (single) {
      for (const segment of single.split(/,(?=[^ =]+=)/)) {
        const nv = segment.split(';')[0].trim();
        if (nv) parts.push(nv);
      }
    }
  }
  return parts.join('; ');
}

/** 合并已有 Cookie 与新响应中的 Cookie */
function mergeCookieHeader(existing: string, headers: Headers): string {
  const incoming = cookieHeaderFromResponse(headers);
  if (!incoming) return existing;
  const map = new Map<string, string>();
  for (const chunk of [existing, incoming]) {
    for (const pair of chunk.split(';')) {
      const trimmed = pair.trim();
      if (!trimmed) continue;
      const eq = trimmed.indexOf('=');
      if (eq <= 0) continue;
      const name = trimmed.slice(0, eq).trim();
      const value = trimmed.slice(eq + 1).trim();
      map.set(name, value);
    }
  }
  return [...map.entries()].map(([k, v]) => `${k}=${v}`).join('; ');
}

/** 从 HTML 页面提取 CSRF token */
function extractCsrfToken(html: string): string {
  const metaMatch = html.match(/<meta\s+name="csrf-token"\s+content="([^"]+)"/i);
  if (metaMatch?.[1]) return metaMatch[1];
  /* 备用：尝试 input hidden */
  const inputMatch = html.match(/<input[^>]+name="_token"[^>]+value="([^"]+)"/i);
  if (inputMatch?.[1]) return inputMatch[1];
  throw new Error('email10min: 未找到 CSRF token');
}

/** 从 HTML 页面提取邮箱地址 */
function extractEmailFromHtml(html: string): string {
  /* 尝试 id="emailAddress" 或 class="emailAddress" 的元素 */
  const idMatch = html.match(/id="emailAddress"[^>]*>([^<]+)</i);
  if (idMatch?.[1]?.includes('@')) return idMatch[1].trim();
  /* 尝试 class 匹配 */
  const clsMatch = html.match(/class="[^"]*email[^"]*"[^>]*>([^<]*@[^<]+)</i);
  if (clsMatch?.[1]) return clsMatch[1].trim();
  /* 尝试 data-email 属性 */
  const dataMatch = html.match(/data-email="([^"]+@[^"]+)"/i);
  if (dataMatch?.[1]) return dataMatch[1].trim();
  /* 尝试 value 属性包含邮箱的 input */
  const valueMatch = html.match(/value="([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})"/);
  if (valueMatch?.[1]) return valueMatch[1].trim();
  /* 尝试 mailbox JSON 数据 */
  const jsonMatch = html.match(/"mailbox"\s*:\s*"([^"]+@[^"]+)"/);
  if (jsonMatch?.[1]) return jsonMatch[1].trim();
  /* 最后尝试：页面中任何合法的临时邮箱地址 */
  const genericMatch = html.match(/([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})/);
  if (genericMatch?.[1]) {
    const addr = genericMatch[1];
    /* 排除常见的非临时邮箱地址 */
    if (!addr.includes('email10min') && !addr.includes('example') && !addr.includes('google')) {
      return addr.trim();
    }
  }
  throw new Error('email10min: 未找到邮箱地址');
}

/** token 编解码：存储 cookie + csrf */
const TOK_PREFIX = 'e10m:';

function encodeToken(cookie: string, csrf: string): string {
  const raw = JSON.stringify({ c: cookie, t: csrf });
  return TOK_PREFIX + Buffer.from(raw, 'utf8').toString('base64');
}

function decodeToken(token: string): { cookie: string; csrf: string } {
  if (!token.startsWith(TOK_PREFIX)) {
    throw new Error('email10min: invalid session token');
  }
  let raw: string;
  try {
    raw = Buffer.from(token.slice(TOK_PREFIX.length), 'base64').toString('utf8');
  } catch {
    throw new Error('email10min: invalid session token');
  }
  const o = JSON.parse(raw) as { c?: string; t?: string };
  const c = (o.c ?? '').trim();
  const t = (o.t ?? '').trim();
  if (!c || !t) throw new Error('email10min: invalid session token');
  return { cookie: c, csrf: t };
}

/**
 * 创建 email10min 临时邮箱
 * 访问页面获取 session cookie、CSRF token 和自动分配的邮箱地址
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const res = await fetchWithTimeout(`${BASE_URL}/zh`, {
    method: 'GET',
    headers: BROWSER_HEADERS,
  });

  if (!res.ok) {
    throw new Error(`email10min: 创建邮箱失败 HTTP ${res.status}`);
  }

  let cookie = cookieHeaderFromResponse(res.headers);
  const html = await res.text();
  const csrf = extractCsrfToken(html);

  const msgRes = await fetchWithTimeout(`${BASE_URL}/messages?${Date.now()}`, {
    method: 'POST',
    headers: { ...AJAX_HEADERS, Cookie: cookie },
    body: `_token=${encodeURIComponent(csrf)}&captcha=`,
  });

  if (!msgRes.ok) {
    throw new Error(`email10min: 获取邮箱失败 HTTP ${msgRes.status}`);
  }

  cookie = mergeCookieHeader(cookie, msgRes.headers);
  const data = await msgRes.json() as { mailbox?: string };
  const email = (data.mailbox ?? '').trim();
  if (!email || !email.includes('@')) {
    throw new Error('email10min: 未获取到邮箱地址');
  }

  return {
    channel: CHANNEL,
    email,
    token: encodeToken(cookie, csrf),
  };
}

/**
 * 获取 email10min 邮件列表
 * 使用 session cookie 和 CSRF token 轮询邮件
 */
export async function getEmails(recipientEmail: string, token: string): Promise<Email[]> {
  const { cookie, csrf } = decodeToken(token);
  const ts = Date.now();

  const response = await fetchWithTimeout(`${BASE_URL}/messages?${ts}`, {
    method: 'POST',
    headers: {
      ...AJAX_HEADERS,
      Cookie: cookie,
    },
    body: `_token=${encodeURIComponent(csrf)}&captcha=`,
  });

  if (!response.ok) {
    throw new Error(`email10min: 获取邮件失败 HTTP ${response.status}`);
  }

  const rawText = await response.text();
  let data: { mailbox?: string; messages?: Array<Record<string, unknown>> };
  try {
    data = JSON.parse(rawText);
  } catch {
    throw new Error(`email10min: 返回非 JSON: ${rawText.slice(0, 120)}`);
  }

  const messages = data.messages || [];
  if (!Array.isArray(messages)) {
    return [];
  }

  return messages.map((raw: any, index: number) =>
    normalizeEmail(
      {
        id: raw.id || raw.message_id || String(index),
        from: raw.from || raw.sender || '',
        to: raw.to || recipientEmail,
        subject: raw.subject || '',
        text: raw.text || raw.body || '',
        html: raw.html || raw.body_html || '',
        date: raw.date || raw.created_at || '',
        isRead: false,
        attachments: raw.attachments || [],
      },
      recipientEmail,
    ),
  );
}

File: sdk/go/provider/email10min.go
package provider

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	http "github.com/bogdanfinn/fhttp"
)

// email10min.com：GET /zh 拿 CSRF + Cookie，POST /messages 获取邮件。

const email10minBase = "https://email10min.com"

var (
	email10minCsrfMetaRe   = regexp.MustCompile(`<meta\s+name="csrf-token"\s+content="([^"]+)"`)
	email10minCsrfInputRe  = regexp.MustCompile(`<input[^>]+name="_token"[^>]+value="([^"]+)"`)
	email10minEmailIdRe    = regexp.MustCompile(`id="emailAddress"[^>]*>([^<]+)`)
	email10minEmailClsRe   = regexp.MustCompile(`class="[^"]*email[^"]*"[^>]*>([^<]*@[^<]+)`)
	email10minDataEmailRe  = regexp.MustCompile(`data-email="([^"]+@[^"]+)"`)
	email10minValueEmailRe = regexp.MustCompile(`value="([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})"`)
	email10minJSONEmailRe  = regexp.MustCompile(`"mailbox"\s*:\s*"([^"]+@[^"]+)"`)
	email10minGenericRe    = regexp.MustCompile(`([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})`)
)

func email10minBrowserHeaders(req *http.Request) {
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8")
	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
	req.Header.Set("Cache-Control", "no-cache")
	req.Header.Set("DNT", "1")
	req.Header.Set("Pragma", "no-cache")
	req.Header.Set("Upgrade-Insecure-Requests", "1")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0")
	req.Header.Set("sec-ch-ua", `"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
}

func email10minAjaxHeaders(req *http.Request) {
	req.Header.Set("Accept", "application/json, text/plain, */*")
	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
	req.Header.Set("Origin", email10minBase)
	req.Header.Set("Referer", email10minBase+"/zh")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0")
	req.Header.Set("X-Requested-With", "XMLHttpRequest")
	req.Header.Set("sec-ch-ua", `"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
	req.Header.Set("sec-fetch-dest", "empty")
	req.Header.Set("sec-fetch-mode", "cors")
	req.Header.Set("sec-fetch-site", "same-origin")
}

type email10minTokenData struct {
	Cookie string `json:"c"`
	CSRF   string `json:"t"`
}

const email10minTokPrefix = "e10m:"

func email10minEncodeToken(cookie, csrf string) string {
	raw, _ := json.Marshal(email10minTokenData{Cookie: cookie, CSRF: csrf})
	return email10minTokPrefix + base64.StdEncoding.EncodeToString(raw)
}

func email10minDecodeToken(token string) (cookie, csrf string, err error) {
	if !strings.HasPrefix(token, email10minTokPrefix) {
		return "", "", fmt.Errorf("email10min: invalid session token")
	}
	decoded, err := base64.StdEncoding.DecodeString(token[len(email10minTokPrefix):])
	if err != nil {
		return "", "", fmt.Errorf("email10min: invalid session token: %w", err)
	}
	var data email10minTokenData
	if err := json.Unmarshal(decoded, &data); err != nil {
		return "", "", fmt.Errorf("email10min: invalid session token: %w", err)
	}
	if data.Cookie == "" || data.CSRF == "" {
		return "", "", fmt.Errorf("email10min: invalid session token (empty fields)")
	}
	return data.Cookie, data.CSRF, nil
}

func email10minExtractCSRF(html string) (string, error) {
	if m := email10minCsrfMetaRe.FindStringSubmatch(html); len(m) > 1 {
		return m[1], nil
	}
	if m := email10minCsrfInputRe.FindStringSubmatch(html); len(m) > 1 {
		return m[1], nil
	}
	return "", fmt.Errorf("email10min: 未找到 CSRF token")
}

func email10minExtractEmail(html string) (string, error) {
	if m := email10minEmailIdRe.FindStringSubmatch(html); len(m) > 1 && strings.Contains(m[1], "@") {
		return strings.TrimSpace(m[1]), nil
	}
	if m := email10minEmailClsRe.FindStringSubmatch(html); len(m) > 1 {
		return strings.TrimSpace(m[1]), nil
	}
	if m := email10minDataEmailRe.FindStringSubmatch(html); len(m) > 1 {
		return strings.TrimSpace(m[1]), nil
	}
	if m := email10minValueEmailRe.FindStringSubmatch(html); len(m) > 1 {
		return strings.TrimSpace(m[1]), nil
	}
	if m := email10minJSONEmailRe.FindStringSubmatch(html); len(m) > 1 {
		return strings.TrimSpace(m[1]), nil
	}
	if m := email10minGenericRe.FindStringSubmatch(html); len(m) > 1 {
		addr := m[1]
		if !strings.Contains(addr, "email10min") && !strings.Contains(addr, "example") && !strings.Contains(addr, "google") {
			return strings.TrimSpace(addr), nil
		}
	}
	return "", fmt.Errorf("email10min: 未找到邮箱地址")
}

// Email10minGenerate 创建 email10min 临时邮箱
func Email10minGenerate() (*CreatedMailbox, error) {
	req, err := http.NewRequest("GET", email10minBase+"/zh", nil)
	if err != nil {
		return nil, err
	}
	email10minBrowserHeaders(req)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "email10min session"); err != nil {
		return nil, err
	}

	cookie := moaktMergeCookies("", resp.Cookies())
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	html := string(raw)

	csrf, err := email10minExtractCSRF(html)
	if err != nil {
		return nil, err
	}
	emailAddr, err := email10minExtractEmail(html)
	if err != nil {
		return nil, err
	}

	token := email10minEncodeToken(cookie, csrf)

	return &CreatedMailbox{
		Channel: "email10min",
		Email:   emailAddr,
		Token:   token,
	}, nil
}

// Email10minGetEmails 获取 email10min 邮件列表
func Email10minGetEmails(email string, token string) ([]NormEmail, error) {
	cookie, csrf, err := email10minDecodeToken(token)
	if err != nil {
		return nil, err
	}

	ts := fmt.Sprintf("%d", time.Now().UnixMilli())
	body := fmt.Sprintf("_token=%s&captcha=", csrf)
	req, err := http.NewRequest("POST", email10minBase+"/messages?"+ts, strings.NewReader(body))
	if err != nil {
		return nil, err
	}
	email10minAjaxHeaders(req)
	req.Header.Set("Cookie", cookie)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "email10min getEmails"); err != nil {
		return nil, err
	}

	rawBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Mailbox  string                   `json:"mailbox"`
		Messages []map[string]interface{} `json:"messages"`
	}
	if err := json.Unmarshal(rawBytes, &data); err != nil {
		return nil, fmt.Errorf("email10min: 返回非 JSON: %s", string(rawBytes[:min(len(rawBytes), 120)]))
	}

	if data.Messages == nil {
		return []NormEmail{}, nil
	}

	emails := make([]NormEmail, 0, len(data.Messages))
	for i, msg := range data.Messages {
		id := ""
		if v, ok := msg["id"]; ok {
			id = fmt.Sprintf("%v", v)
		} else if v, ok := msg["message_id"]; ok {
			id = fmt.Sprintf("%v", v)
		} else {
			id = fmt.Sprintf("%d", i)
		}
		from, _ := msg["from"].(string)
		if from == "" {
			if v, ok := msg["sender"]; ok {
				from = fmt.Sprintf("%v", v)
			}
		}
		to, _ := msg["to"].(string)
		if to == "" {
			to = email
		}
		subject, _ := msg["subject"].(string)
		text, _ := msg["text"].(string)
		if text == "" {
			text, _ = msg["body"].(string)
		}
		htmlContent, _ := msg["html"].(string)
		if htmlContent == "" {
			htmlContent, _ = msg["body_html"].(string)
		}
		date, _ := msg["date"].(string)
		if date == "" {
			date, _ = msg["created_at"].(string)
		}

		flat := map[string]interface{}{
			"id":      id,
			"from":    from,
			"to":      to,
			"subject": subject,
			"text":    text,
			"html":    htmlContent,
			"date":    date,
			"isRead":  false,
		}
		emails = append(emails, NormalizeMap(flat, email))
	}
	return emails, nil
}
